### PR TITLE
Refactor requests

### PR DESCRIPTION
**Pull request by leplatrem**
* [x] Wait for #93 to be merged
* [x] Refactor
* [x] Integration tests pass
* [x] Make unit tests pass

### CODE REVIEW

**Comment by n1k0** [comment]
Hmm "This branch has conflicts that must be resolved". Anyway, the patch looks good, r+ with green builds and conflicts resolved ;)

**Comment by n1k0** [comment]
Hmm scratch that, wrong PR.

**Comment by leplatrem** [comment]
This is now ready to be reviewed.

Note that integration tests remain unchanged ;)

File: src/base.js [hunk 1]
```diff
@@ -396,8 +396,16 @@ export default class KintoClientBase {
    * @return {Promise<Object, Error>}
    */
   createBucket(bucketName, options={}) {
+    if (!bucketName) {
+      throw new Error("A bucket name is required.");
```
Comment (leplatrem) [inline src/base.js]: nit: bucket **name** vs. bucket **id**

File: test/api_test.js [hunk 2]
```diff
@@ -575,14 +575,21 @@ describe("KintoClient", () => {
   /** @test {KintoClient#createBucket} */
   describe("#createBucket", () => {
     beforeEach(() => {
-      sandbox.stub(requests, "createBucket");
+      sandbox.stub(requests, "createRequest");
       sandbox.stub(api, "execute").returns(Promise.resolve());
     });
 
+    it("should throw if bucketName is missing", () => {
```
Comment (leplatrem) [inline test/api_test.js]: nit: bucket id

File: src/base.js [hunk 3]
```diff
@@ -411,8 +419,14 @@ export default class KintoClientBase {
...
   deleteBucket(bucket, options={}) {
+    const bucketObj = toDataBody(bucket);
+    if (!bucketObj.id) {
+      throw new Error("A bucket id is required.");
+    }
+    const path = endpoint("bucket", bucketObj.id);
+    options = { last_modified: bucket.last_modified, ...options };
     const reqOptions = this._getRequestOptions(options);
-    return this.execute(requests.deleteBucket(toDataBody(bucket), reqOptions));
+    return this.execute(requests.deleteRequest(path, reqOptions));
```
Comment (n1k0) [inline src/base.js]: That's nice :+1:

File: src/bucket.js [hunk 4]
```diff
@@ -99,8 +99,22 @@ export default class Bucket {
...
+    if (typeof data !== "object") {
+      throw new Error("A bucket object is required.");
+    }
+
+    const bucket = {...data, id: this.name};
+
+    // For default bucket, we need to drop the id from the data object.
+    const bucketId = bucket.id;
+    if (bucket.id === "default") {
+      delete bucket.id;
```
Comment (n1k0) [inline src/bucket.js]: Didn't I see a server patch landed about that lately?

File: src/bucket.js [hunk 5]
```diff
@@ -131,7 +145,11 @@ export default class Bucket {
    */
   createCollection(id, options={}) {
     const reqOptions = this._bucketOptions(options);
-    const request = requests.createCollection(id, reqOptions);
+    const { permissions, data={} } = reqOptions;
+    data.id = id;
+    const path = id ? endpoint("collection", this.name, id) :
+                      endpoint("collections", this.name);
```
Comment (n1k0) [inline src/bucket.js]: In theory simply `const path = endpoint("collection", this.name, id);` should get us covered.

File: src/bucket.js [hunk 6]
```diff
@@ -145,8 +163,15 @@ export default class Bucket {
    * @return {Promise<Object, Error>}
    */
   deleteCollection(collection, options={}) {
+    const collectionObj = toDataBody(collection);
+    if (!collectionObj.id) {
+      throw new Error("A collection id is required.");
+    }
+
+    options = { last_modified: collectionObj.last_modified, ...options };
```
Comment (n1k0) [inline src/bucket.js]: It's probably worth destructuring `collectionObj` above, eg. `const {id, last_modified} = collectionObj;`

File: src/bucket.js [hunk 7]
```diff
@@ -177,10 +202,14 @@ export default class Bucket {
    * @return {Promise<Object, Error>}
    */
   setPermissions(permissions, options={}) {
-    return this.client.execute(requests.updateBucket({
-      id: this.name,
-      last_modified: options.last_modified
-    }, {...this._bucketOptions(options), permissions}));
+    if (typeof permissions !== "object") {
```
Comment (n1k0) [inline src/bucket.js]: Beware of this test for testing for actual objects, eg `null` and `array` are both passing this test (yeah, I know). 

Even if not blocking in this specific case, here's a simple helper you could probably use https://github.com/Kinto/kinto-admin/blob/97a61ec68a9608d48de4267072263f2fd24a37a6/scripts/utils.js#L21-L23

File: src/bucket.js [hunk 8]
```diff
@@ -177,10 +202,14 @@ export default class Bucket {
...
+      throw new Error("A permissions object is required.");
+    }
+    const path = endpoint("bucket", this.name);
+    const reqOptions = {...this._bucketOptions(options)};
+    const data = { last_modified: options.last_modified };
```
Comment (n1k0) [inline src/bucket.js]: I'm now probably a little too much addicted to object destructuring, but maybe:

```js
const {last_modified} = options;
const data = {last_modified};
```

File: src/collection.js [hunk 9]
```diff
@@ -149,7 +156,11 @@ export default class Collection {
    */
   createRecord(record, options={}) {
     const reqOptions = this._collOptions(options);
-    const request = requests.createRecord(this.name, record, reqOptions);
+    const { permissions } = reqOptions;
+    // XXX throw if bucket is undefined
+    const path = record.id ? endpoint("record", this.bucket.name, this.name, record.id) :
+                             endpoint("records", this.bucket.name, this.name);
```
Comment (n1k0) [inline src/collection.js]: Same remark as previously, when `id` is `undefined` the `endpoint` function will see it as a non-passed arg.

File: src/requests.js [hunk 10]
```diff
@@ -78,173 +84,3 @@ function _delete(path, options={}) {
...
-    throw new Error("A record id is required.");
-  }
-  const { bucket } = {
-    ...requestDefaults,
-    ...options
-  };
-  options = { last_modified: record.last_modified, ...options };
-  const path = endpoint("record", bucket, collName, record.id);
-  return _delete(path, options);
-}
```
Comment (n1k0) [inline src/requests.js]: So long silly functions :dancer:

File: src/utils.js [hunk 11]
```diff
@@ -183,7 +183,3 @@ export function nobatch(message) {
     };
   };
 }
-
-export function removeUndefined(obj) {
-  return JSON.parse(JSON.stringify(obj));
-}
```
Comment (n1k0) [inline src/utils.js]: I won't miss ya :kiss:

File: test/bucket_test.js [hunk 12]
```diff
@@ -206,22 +209,21 @@ describe("Bucket", () => {
...
 
-        sinon.assert.calledWithMatch(requests.createCollection, "foo", {
-          bucket: "blog",
-          headers: {Foo: "Bar", Baz: "Qux"},
-          safe: true,
-        });
+        sinon.assert.calledWithMatch(requests.createRequest, "/buckets/blog/collections/foo", {
+          data: {id: "foo"},
+          permissions: undefined
+        }, { headers: {Foo: "Bar", Baz: "Qux"}, safe: true });
```
Comment (n1k0) [inline test/bucket_test.js]: Maybe spread the options on several lines here and elsewhere for consistency and as it really improves legibility to me.

File: test/collection_test.js [hunk 13]
```diff
@@ -167,23 +162,27 @@ describe("Collection", () => {
...
 
     it("should accept a safe option", () => {
-      sandbox.stub(requests, "createRecord");
+      sandbox.stub(requests, "createRequest");
 
       coll.createRecord(record, {safe: true});
 
-      sinon.assert.calledWithMatch(requests.createRecord, "posts", record, {
-        safe: true
+      sinon.assert.calledWithMatch(requests.createRequest, "/buckets/blog/collections/posts/records", {
```
Comment (n1k0) [inline test/collection_test.js]: To be noted we're heavily relying on `calledWithMatch` while I'm pretty sure most of our assertions should just pass using `calledWithExactly`.

File: test/collection_test.js [hunk 14]
```diff
@@ -210,49 +209,47 @@ describe("Collection", () => {
       sandbox.stub(client, "execute").returns(Promise.resolve({data: 1}));
     });
 
+    it("should throw if record is not an object", () => {
+      expect(() => coll.updateRecord(2)).to.Throw(Error, /required/);
+    });
+
+    it("should throw if id is missing", () => {
+      expect(() => coll.updateRecord({})).to.Throw(Error, /required/);
```
Comment (n1k0) [inline test/collection_test.js]: Could we have more precise test strings to reflect we're actually getting the expected messages?

File: test/collection_test.js [hunk 15]
```diff
@@ -262,38 +259,36 @@ describe("Collection", () => {
...
     beforeEach(() => {
+      sandbox.stub(requests, "deleteRequest");
       sandbox.stub(client, "execute").returns(Promise.resolve({data: 1}));
     });
 
-    it("should delete a record", () => {
-      sandbox.stub(requests, "deleteRecord");
+    it("should throw if id is missing", () => {
+      expect(() => coll.deleteRecord({}))
+        .to.Throw(Error, /required/);
```
Comment (n1k0) [inline test/collection_test.js]: Ditto

**Comment by n1k0** [comment]
AMAZING WORK, this is so much better, thanks for tackling this. A few minor remark, r+wc

File: src/bucket.js (hunk 4 above)
Comment (leplatrem) [inline src/bucket.js]: yup, do we want it for older server versions (ie. stage/prod at Moz) ?

Comment (n1k0) [inline src/bucket.js]: ah the good 'ol question with backward compat and version support. tbh, idk. was it just me, i'd just support master :D

Comment (Natim) [inline src/bucket.js]: I think it doesn't hurt to keep it but let's add a comment maybe (For Kinto < 3.0)

Comment (n1k0) [inline src/bucket.js]: Yup agreed

File: src/endpoint.js [hunk 16]
```diff
@@ -5,12 +5,9 @@
...
   root:                   () => "/",
   batch:                  () => "/batch",
-  buckets:                () => "/buckets",
-  bucket:           (bucket) => `/buckets/${bucket}`,
-  collections:      (bucket) => `${ENDPOINTS.bucket(bucket)}/collections`,
-  collection: (bucket, coll) => `${ENDPOINTS.bucket(bucket)}/collections/${coll}`,
-  records:    (bucket, coll) => `${ENDPOINTS.collection(bucket, coll)}/records`,
-  record: (bucket, coll, id) => `${ENDPOINTS.records(bucket, coll)}/${id}`,
+  bucket:           (bucket) => "/buckets" + (bucket ? `/${bucket}` : ""),
+  collection: (bucket, coll) => `${ENDPOINTS.bucket(bucket)}/collections` + (coll ? `/${coll}` : ""),
```
Comment (n1k0) [inline src/endpoint.js]: Not sure if much better but shorter:

```js
collection: (bucket, coll) => `${ENDPOINTS.bucket(bucket)}/collections${coll ? `/${coll}` : ""}`,
```

**Comment by n1k0** [comment]
r+ o/

**Pull request merged**